torch/utils/data/datapipes/map/grouping.py

```diff
@@ -0,0 +1,89 @@
+import operator
+
+from functools import reduce
+from torch.utils.data import MapDataPipe, functional_datapipe, DataChunk
+from typing import List, TypeVar, Union, Optional
+
+
+T = TypeVar('T')
+
+
+@functional_datapipe('unbatch')
+class UnBatcherMapDataPipe(MapDataPipe[T]):
+    r""" :class:`UnBatcherMapDataPipe`.
+
+    Map DataPipe to undo batching of data. In other words, it flattens the data up to the specified level
+    within a batched DataPipe.
+
+    Args:
+        datapipe: Iterable DataPipe being un-batched
+        unbatch_level: Defaults to `1` (only flattening the top level). If set to `2`, it will flatten the top 2 levels,
+            and `-1` will flatten the entire DataPipe.
+    """
+
+    def __init__(self,
+                 datapipe: MapDataPipe[DataChunk],
+                 unbatch_level: int = 1):
+        if (not isinstance(unbatch_level, int)) or unbatch_level < -1:
+            raise RuntimeError(f"unbatch_level must be an integer > -1, got {unbatch_level} instead.")
+        self.datapipe = datapipe
+        self.deepest_level = self._get_deepest_level()
+        self.unbatch_level = unbatch_level if unbatch_level != -1 else self.deepest_level
+        self.length: Optional[int] = None
+
+    def _get_batch_sizes(self) -> List[int]:
+        sizes = [len(self.datapipe)]
+        curr = self.datapipe.__getitem__(0)
+        while isinstance(curr, list) or isinstance(curr, DataChunk):
+            sizes.append(len(curr))
+            curr = curr[0]
+        return sizes
+
+    def _get_deepest_level(self) -> int:
+        level = 0
+        curr: Union[MapDataPipe, DataChunk, List] = self.datapipe
+        try:
+            while isinstance(curr[0], list) or isinstance(curr[0], DataChunk):
+                curr = curr[0]
+                level += 1
+        except IndexError:
+            pass
+        return level
+
+    def __getitem__(self, index) -> T:
```

Comment:
The complicated logic is here. This part maps the new, given index to the elements within input, batched DataPipe.
The logic assumes that the input DataPipe is batched properly with full batches (except maybe the last batch), otherwise, an exception is raised.